### PR TITLE
Revamp Bits strategy and implement Bitstruct strategy

### PR DESCRIPTION
**Pull request by jsn1993**
Now I think we should create strategies in a generator-like style. At strategy construction time, our bitwidth/bitstruct types are already fixed, which means all the arguments won't change during runtime. As a result, we should be able to capture them in closure and return a strategy with minimum metadata access like object dict access or creating types.

Based on these, I reimplemented Bits strategy and added custom min_value/max_value support since there can be cases where the full range of value doesn't make sense. I also write a bitstruct strategy based on recursive construction of strategy list, and the real strategy is very succinct and simple.

For complete functionality, I think I still need to implement setting min/max value limit from the top-level bitstruct using a trie-like data structure. I open the PR early to get some feedback on style.

### CODE REVIEW

**Comment by jsn1993** [comment]
Setting manual min/max limit already implemented. Feature complete now! @yo96 take a look when you have time. @icyxiao96 you can take a look if you want

**Comment by cbatten** [comment]
This is a great idea!

File: pymtl3/datatypes/strategies.py [hunk 1]
```diff
@@ -4,30 +4,114 @@
...
+
+  return strategy_list() # RETURN A STRATEGY INSTEAD OF FUNCTION
+
+#-------------------------------------------------------------------------
+# strategies.bitstruct
+#-------------------------------------------------------------------------
+# Return the SearchStrategy for bitstruct type T with the support of
+# dictionary-based min/max value limit
+
+def bitstruct( T, limit_dict=None ):
```
Comment (yo96) [inline pymtl3/datatypes/strategies.py]: bitstructs

File: pymtl3/datatypes/strategies.py [hunk 2]
```diff
@@ -4,30 +4,114 @@
...
+
+  return strategy_bits() # RETURN A STRATEGY INSTEAD OF FUNCTION
+
+#-------------------------------------------------------------------------
+# strategies.bitslist
+#-------------------------------------------------------------------------
+# Return the SearchStrategy for a list of Bits with the support of
+# dictionary based min/max value limit
+
+def bitslist( types, limit_dict=None ):
```
Comment (yo96) [inline pymtl3/datatypes/strategies.py]: bitslists

Comment (cbatten) [reply pymtl3/datatypes/strategies.py]: I think it should be bitslist ... it is a list of bits not multiple lists of bits?

File: pymtl3/datatypes/test/strategies_test.py [hunk 3]
```diff
@@ -37,4 +40,172 @@ def test_signed( nbits ):
...
+    assert 0xf0 <= bs.p2.y <= 0xff
+    print( bs )
+
+  print("-"*10,NestedPoint,"-"*10)
+  actual_test()
+
+def test_nnested_point_limited():
+  limit_dict = {
+    'p1': {
+      'x': (0xe0,0xef),
```
Comment (yo96) [inline pymtl3/datatypes/test/strategies_test.py]: we should support user defined strategies here too

File: pymtl3/datatypes/strategies.py (hunk 2 above)
Comment (yo96) [reply pymtl3/datatypes/strategies.py]: well in hypothesis it's also st.lists not list, because it's a strategy that generate different lists.

Comment (cbatten) [reply pymtl3/datatypes/strategies.py]: Ah ha -- I stand corrected :) we should definitely follow the hypothesis naming conventions.

File: pymtl3/datatypes/test/strategies_test.py (hunk 3 above)
Comment (jsn1993) [reply pymtl3/datatypes/test/strategies_test.py]: Done. Check the latest tests!

File: pymtl3/datatypes/test/strategies_test.py [hunk 4]
```diff
@@ -25,10 +26,12 @@ def test_unsigned( nbits ):
   def actual_test( bits ):
     assert bits.nbits == nbits
     print( bits, bits.uint() )
+
+  print("-"*10,nbits,"-"*10)
```
Comment (yo96) [inline pymtl3/datatypes/test/strategies_test.py]: maybe we can get rid of these print?

**Pull request merged**